Poetry is to be

MattiasMTS opened the pull request:
# What has changed:
For the reviewer: Feel free to add any author(s) you seem fit in the `pyproject.toml` file. Note that any new version of `impedance.py` has to be reflected by updating the `version` number in the `pyproject.toml` file. 

1. Poetry has been added as python package manager. During the development of this, I saw that the dependencies needed python 3.7 to be used as the lowest python interpreter. I believe it was `flake8` that was complaining about it. Nevertheless, if you wish to use the latest `poetry` (1.2) then 3.7 is also required. Solves #215 

2. Furthermore, I also updated the CI workflow to re-use `impedance.py` own workflow (defined in `.github/actions/build_impedance/action.yml`). Just makes the building a bit cleaner. Additionally, a `release.yml` workflow is triggered on `release`. Quite tricky to test this but based on the [docs](https://python-poetry.org/docs/cli/#publish) this should be the correct syntax for this. 

3. Added some more docs on how to use poetry as your virtual environment + package manager in the `CONTRIBUTING.md` file. 

Let me know if there is any uncertainty or questions. Happy to iterate here!

One bottleneck I came across when adding the dependencies listed in the requirement.txt file was that it took suuuuuper long to add the `altair` package (after adding all the others)
![image](https://user-images.githubusercontent.com/86059470/195605847-009e72e1-d570-4aa2-af60-00260ce9f1db.png)

Makes me a bit sad that it takes so long time but to update the packages and keep it up-to-date you can run `poetry update`, which might take some time, unfortunately. 

Let me know if you want `pre-commit` to be added as well in subsequential PRs.

mdmurbach commented on pyproject.toml:
```diff
@@ -0,0 +1,24 @@
+name = "impedance.py"
+version = "1.4.1"
+description = "A Python package for working with electrochemical impedance data."
+authors = ["Your Name <you@example.com>"]  # TODO add your name here
```
```suggestion
authors = ["impedance.py developers <matt.murbach@gmail.com>"]
```

mdmurbach reviewed (commented):
Thanks for putting this PR together. I've been playing around with poetry a bit and I have one question about the .lock file

In particular, I get this warning about the .lock file not being consistent with the pyproject.toml when I go to install using poetry

![image](https://user-images.githubusercontent.com/9369020/197090066-1e5a5909-e0bf-48fd-89f1-ed980cf4cc48.png)

Running `poetry lock --check` confirms this

![image](https://user-images.githubusercontent.com/9369020/197090565-ad593a88-039b-45b0-93a4-b70440ad52d1.png)

My question is whether this is because we are on different operating systems (maybe Python versions) or if the file just got out of sync and you just need to run poetry lock again to update what's in the branch?

MattiasMTS commented:
Ah yes, I manually changed the supported python versions in the `pyproject.toml` and forgot to execute `poetry lock` afterwards. I will do a `poetry lock` commit, then you shouldn't see this warning messages anymore. 

Furthermore, I believe it makes more sense to do the tests refactoring I mentioned in #228, in this PR. Since I have already split the dev and user dependencies here. Is this ok with you @mdmurbach ?

mdmurbach commented:
Hi @MattiasMTS, thanks for the pyproject.toml update 👍 

Sorry I've been so slow on all of these discussions (family has traded off round robin being sick with the flu the last few weeks...), but I think this is looking good! I agree that the refactor of tests probably makes sense in this PR. For testing out the Actions release workflow, I'm thinking that I'll test it once it's merged and make any changes (don't think making a release on your branch will work since there's no pypi secret)

> Happy to know that you managed to play around with poetry, what do you think so far?

Poetry is cool (and it's pretty wild that I've spent so long without "real" Python package management). I think I like the combo of conda for Python version + non-Python packages and Poetry for the project dependencies. Very nice how poetry seems to do a lot without a lot of overhead.

MattiasMTS commented:
So, linting is failing because forgot to add flake8. I will add this. 

Regarding why python 3.10 is failing is because of some inconsistencies with CPython and numpy. I tried to find the best version for numpy and 3.10 by using poetry update, but as mentioned it takes super long time. 

The issue why poetry takes long is because of the Altair package. Therefore, my question is are we using this for something specific as of now? If it would be possible to remove its dependency

BGerwe commented:
> The issue why poetry takes long is because of the Altair package. Therefore, my question is are we using this for something specific as of now? If it would be possible to remove its dependency

We do use altair for interactive plotting. Removing that would be a potentially breaking change for some users, so I don't think we should remove it without giving adequate deprecation notice.

momoson commented:
After tinkering a little bit with poetry it seems like the python version specification `python = ">=3.7,<3.11"` is the culprit. 
By changing the python version specification to `python ="^3.8"`, the required time for `poetry update` was reduced to ~6 seconds. Even after clearing poetry's package cache, update time only increased to ~30 seconds.
It seems like Poetry tries to find suitable dependencies for each specified python version and that's particular hard for Python version 3.7.

Python version 3.7 has now already reached [end-of-life](https://peps.python.org/pep-0537/). So it can IMHO be removed from the supported python versions without further ado.

@MattiasMTS, could you add the python version change to this PR? Do you still plan on removing the `requirements.txt` file? Keeping it seems somewhat redundant. Apart from that, is there anything else keeping this from being merged?

MattiasMTS commented:
> After tinkering a little bit with poetry it seems like the python version specification `python = ">=3.7,<3.11"` is the culprit. By changing the python version specification to `python ="^3.8"`, the required time for `poetry update` was reduced to ~6 seconds. Even after clearing poetry's package cache, update time only increased to ~30 seconds. It seems like Poetry tries to find suitable dependencies for each specified python version and that's particular hard for Python version 3.7.
> 
> Python version 3.7 has now already reached [end-of-life](https://peps.python.org/pep-0537/). So it can IMHO be removed from the supported python versions without further ado.
> 
> @MattiasMTS, could you add the python version change to this PR? Do you still plan on removing the `requirements.txt` file? Keeping it seems somewhat redundant. Apart from that, is there anything else keeping this from being merged? If you are not interested anymore in getting this merge, I would try to rebase your commits onto the newest main, introduce those discussed minor changes and get this PR merged.

Thanks for taking a look at this @momoson and sorry @mdmurbach that I have left this in stale. I'll take a look at what needs to be changed tomorrow or this weekend and get back to you ✨

MattiasMTS commented:
## NB
Rebased with respect to main now. I have added some essential changes (let me know if you disagree and I can revert). The changes are:

- move pylama configuration to the pyproject.toml file and added it as a dev dependency
- added flake8 as dev dependency
- refactored the ci.release worker to use the build_impedance re-usable workflow by adding its dependency as "release" in the pyproject.toml file (this will make poetry only install this as its dependency).
- Dropped support for 3.7 and update it in places accordingly
- Used the `importlib.metadata` library to extract the version of the impedance library (since this is mostly supported for python +3.8, thought it would be a good idea). The version of the library is now defined from the pyproject.toml attribute. 
- Removed the requirements.txt file. Let me know if you still want this for users which are using e.g. `venv` and still do `pip install -r requirements.txt` but just using `python venv`. 
- Fixed some flake8 issues in the tests by ignoring rule `# noqa: F811`


## Questions
Since this is sort of a deprecation PR (because we are dropping support for python 3.7) we need to bump the impedance.__version__ accordingly. Do you want to bump it to 1.8.0? 

Secondly, it is sort of recommended for "library" repository to skip the git track of the `poetry.lock` file. Read more about here https://python-poetry.org/docs/basic-usage/#committing-your-poetrylock-file-to-version-control and let me know your stance on this.

Once this PR is merged I will handle any rebase needed for the #228 PR. The changes there will be minimum but I have some new pre-commit-hooks I want to add as well.

momoson commented:
Thank you for the fast and clean rebase @MattiasMTS!

The `setup.py` file is flawed now, anyway, hence I vote for removing it. When no version of `impedance.py` is installed on the system, the first line of `setup.py` will import `__version__` to be `Could not find impedance installed`, leading to further errors. However, when `impedance` *is* installed, `__version__` will be set to the version of the installed `impedance` package, rather than the actual version of the package setuptools is building right now.
This behavior is also what confuses me most about `importlib.metadata`. When having a local folder `impedance` with version 2.0 and an installed package called `impedance` with version 1.8, `import impedance; print(impedance.__version__)` *should* return 2.0. Instead, it returns 1.8.

An alternative approach to `importlib.metadata` might be the [`poetry-bumpversion` plugin](https://pypi.org/project/poetry-bumpversion/). The version number would be stored both in `__init__.py` and `pyproject.toml`, however, to bump versions only one call to `poetry version bump` is required.

As far as I could see when stumbling through the code of `gh-action-pypi-publish`, there's also no need for the `setup.py`. It just uploads the contents of `dist/*` to PyPI using `twine` which has no dependency on `setup.py`. So independent from the decision `poetry build` vs. `gh-action-pypi-publish` action, `setup.py` can be dropped.

Regarding the movement of metadata from `setup.py` to `pyproject.toml`: Maybe try to upload it to https://test.pypi.org/ first to see differences, if any.

MattiasMTS commented:
> The setup.py file is flawed now, anyway, hence I vote for removing it. When no version of impedance.py is installed on the system, the first line of setup.py will import __version__ to be Could not find impedance installed, leading to further errors. However, when impedance is installed, __version__ will be set to the version of the installed impedance package, rather than the actual version of the package setuptools is building right now.
This behavior is also what confuses me most about importlib.metadata. When having a local folder impedance with version 2.0 and an installed package called impedance with version 1.8, import impedance; print(impedance.__version__) should return 2.0. Instead, it returns 1.8.

> An alternative approach to importlib.metadata might be the [poetry-bumpversion plugin](https://pypi.org/project/poetry-bumpversion/). The version number would be stored both in __init__.py and pyproject.toml, however, to bump versions only one call to poetry version bump is required.

I believe the issue with having a local development folder _and_ a pip installed version of the `impedance` library comes down to how python prioritize the `PATH` of the lib. `importlib.metadata` will by default only look in your `pip` list of available versions. If the local development version is set to a different one than the one you've installed; I'd argue it is doing the right thing. 

I don't see how this is an issue rather than it might confuse developers. Regardless, if you've setup your virtual environment and `PATH` correctly, you should be able to deduce which is which, i.e. always prioritize the local path folder rather than pip installed one. A fallback to this is to read the `pyproject.version` during the exception. Feel free to add this if you think it is the right way forward.

> As far as I could see when stumbling through the code of gh-action-pypi-publish, there's also no need for the setup.py. It just uploads the contents of dist/* to PyPI using twine which has no dependency on setup.py. So independent from the decision poetry build vs. gh-action-pypi-publish action, setup.py can be dropped.

I have not utilized `poetry build` which defaults to the bin folder. We could utilize the `poetry release` cmd for this as well and skip the `gh-action-pypi-publish` workflow dependency. But this could be done in a separate PR since this PR comes with enough new changes. 

I encourage anyone who is admin of the repo to try and upload this branch to pypi and do some trial-and-error to see how much is changed by moving the metadata from `setup.py` to `pyproject.toml`. Feel free to commit a change of dropping the `setup.py` file whenever you're happy with the switch.


I'll be off tomorrow and today. So, feel free to make any changes you seem fit to this PR if you need it merged ASAP 😋

momoson commented:
I've uploaded the package to testpypi: https://test.pypi.org/project/impedance-py/
Metadata looks good to me. However, the package name in `pyproject.toml` must be changed, otherwise it will be uploaded to a different PyPi project.

`poetry build` seems to default to the `dist` folder now, making it compatible with the `ghttps://test.pypi.org/project/impedance-py/` action.

Unfortunately I cannot add commits myself, hence here's a series of patch files. @mdmurbach just apply the ones you like.
[0006-maint-Bump-version-to-1.8.0.txt](https://github.com/ECSHackWeek/impedance.py/files/12062765/0006-maint-Bump-version-to-1.8.0.txt)
[0004-maint-Update-altair-dependency-in-README.txt](https://github.com/ECSHackWeek/impedance.py/files/12062766/0004-maint-Update-altair-dependency-in-README.txt)
[0003-maint-Remove-.py-suffix-from-poetry-package-name.txt](https://github.com/ECSHackWeek/impedance.py/files/12062767/0003-maint-Remove-.py-suffix-from-poetry-package-name.txt)
[0002-janitor-Remove-setup.py.txt](https://github.com/ECSHackWeek/impedance.py/files/12062768/0002-janitor-Remove-setup.py.txt)

MattiasMTS commented:
Thanks for the patch files @momoson ! 

The only decision left is if we want to git track the `poetry.lock` file (see https://python-poetry.org/docs/basic-usage/#committing-your-poetrylock-file-to-version-control)

@mdmurbach Ready for re-review, I guess 😋
